### PR TITLE
[MLPerf GPT3] Data pipeline, Eval step

### PR DESCRIPTION
**Pull request by ZhiyuLi-goog**
As titled, Data pipeline, Eval step and compiliance log for MLPerf

* eval step
* new c4_mlperf dataset for mlperf
* mllog i.e. compiliance log for MLPerf
* switch to iterator class `MultiHostDataLoadIterator` to support for loop in evaluation step
  - `reset` iterator at each evaluation step
  - iterator class supports for loop naturally

### CODE REVIEW

**Review by rwitten — CHANGES_REQUESTED** [review]
I believe there are CR's missing from this so temporarily removing myself as assignee.

**Comment by ZhiyuLi-goog** [comment]
> I believe there are CR's missing from this so temporarily removing myself as assignee.

Hi, @rwitten You are right. 
This CR is a follow-up one which depends on [the first change](https://github.com/google/maxtext/pull/293). We can go back to this one later.

**Comment by ZhiyuLi-goog** [comment]
The first PR https://github.com/google/maxtext/pull/293 has been merged in and this CL is ready for review.

File: MaxText/multihost_dataloading.py [hunk 1]
```diff
@@ -105,3 +105,21 @@ def get_next_batch_sharded(
   input_gdas = jtu.tree_map_with_path(partial(_form_global_array, global_mesh = global_mesh), local_data)
 
   return input_gdas
+
+
+class MultiHostDataLoadIterator:
```
Comment (rwitten) [inline MaxText/multihost_dataloading.py:110]: @aireenmei be aware.

@ZhiyuLi-goog looks great!

File: MaxText/pyconfig.py [hunk 2]
```diff
@@ -149,13 +149,21 @@ def __init__(self, argv: list[str], **kwargs):
...
     keys_from_env_and_command_line = self._update_from_env_and_command_line(raw_keys, raw_data_from_yaml, argv, **kwargs)
     max_logging.log(
         f"Updating keys from env and command line: {keys_from_env_and_command_line}")
+
     keys_from_model = _HyperParameters.update_model_vars(raw_keys)
     max_logging.log(f"Updating keys from model: {keys_from_model}")
     validate_no_keys_overwritten_twice(keys_from_env_and_command_line, keys_from_model)
 
+    # We initialize the jax distributed system here because it must be done before device backend is initialized.
+    if raw_keys["enable_checkpointing"] and raw_keys["async_checkpointing"] and raw_keys["compile_topology_num_slices"]==-1:
```
Comment (rwitten) [inline MaxText/pyconfig.py:162]: @gobbleturk I'm really confused that this is spreading but we need to figure out how to make it stop. Should the three of us get together and discuss this?

File: MaxText/train.py [hunk 3]
```diff
@@ -196,33 +197,55 @@ def train_step(model, config, state, data, dropout_rng):
...
+      data[k] = v[:config.global_batch_size_to_train_on,:]
 
-  def loss_fn(params):
+  def loss_fn(params, is_train=True):
     logits, intermediate_outputs = model.apply({'params': params},
                          data['inputs'],
                          data['inputs_position'],
                          decoder_segment_ids=data['inputs_segmentation'],
-                         enable_dropout=config.enable_dropout,
+                         enable_dropout=config.enable_dropout if is_train else False,
```
Comment (rwitten) [inline MaxText/train.py:209]: Why? Can't you set Dropout= 0?

File: MaxText/train.py [hunk 4]
```diff
@@ -301,16 +324,30 @@ def train_loop(config, state=None):
   Returns:
   """
   ( init_rng, writer, checkpoint_manager, state_mesh_annotations, model,
-  mesh, learning_rate_schedule, data_iterator, state) = setup_train_loop(config)
+  mesh, learning_rate_schedule, data_iterator, eval_data_iterator, state) = setup_train_loop(config)
 
-  functional_train, in_shard, out_shard, static_argnums, donate_argnums = maxtext_utils.get_functional_train_with_signature(
+  # pylint: disable=line-too-long
+  functional_train, in_shard, out_shard, static_argnums_train, donate_argnums_train = maxtext_utils.get_functional_train_with_signature(
```
Comment (rwitten) [inline MaxText/train.py:330]: Not good to have a "functional_train_with_signature" that takes "is_train". Happy to discuss solutions or for you to propose one!

File: MaxText/train.py [hunk 5]
```diff
@@ -345,6 +391,10 @@ def train_loop(config, state=None):
 
   example_batch = None
   last_step_completion = datetime.datetime.now()
+  if config.enable_mllog:
```
Comment (rwitten) [inline MaxText/train.py:394]: probably wrap this in a maxtext specific version to preserve readability in train.py.

File: MaxText/train.py [hunk 6]
```diff
@@ -371,6 +421,23 @@ def train_loop(config, state=None):
...
+          _, eval_metrics = p_eval_step(
+            state, eval_batch, nextrng
+          )
+        cumulative_eval_metrics['total_loss'] += float(eval_metrics['scalar']['evaluation/total_loss'])
+        cumulative_eval_metrics['total_weights'] += float(eval_metrics['scalar']['evaluation/total_weights'])
+      eval_loss = cumulative_eval_metrics['total_loss'] / (cumulative_eval_metrics['total_weights'] + EPS)
+      max_logging.log(f"average loss after {step=}: {eval_loss=}, total_weights={cumulative_eval_metrics['total_weights']}")
+      if config.enable_mllog:
+        mllog_utils.early_stop_check(config, step, eval_loss)
+      if eval_loss <= config.target_eval_loss:
```
Comment (rwitten) [inline MaxText/train.py:438]: Shouldn't we log here?

File: MaxText/train.py (hunk 3 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py:209]: Hi, @rwitten 

`Dropout = 0`(`dropout_rate=0`) is equivalent to `enable_dropout = False`, while `model` allows us to pass `enable_dropout` only and `dropout_rate` is fixed in config. Considering we enable dropout with a valid `dropout_rate` in training while disabling it evaluation.

Looking into the code in depth, `deterministic` is simply designed to disable dropout in evaluation:
```
# flax https://flax.readthedocs.io/en/latest/_modules/flax/linen/stochastic.html#Dropout
x = nn.Dropout(0.5, deterministic=not train)(x)

# ours
deterministic=not enable_dropout,
...
nn.Dropout(
        rate=cfg.dropout_rate, broadcast_dims=(-2,)
    )(x, deterministic=deterministic)
```

File: MaxText/train.py (hunk 4 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py:330]: Thank you @rwitten for the comments.

Instead of fusing train and evaluation in the same `functional_train_with_signature`, I am considering to add a `functional_eval_with_signature` in addition to `functional_train_with_signature`, what do you think?

```
def get_functional_eval_step(train_step, model, config):
  # train and eval step share the same api for simplicity:
  #   train:  new_state, metrics = train_step(state, data, is_train=True)
  #   eval:  _, metrics = train_step(state, data, is_train=False)
  return functools.partial(train_step, model, config, is_train=False)

def get_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config):
  ...

# add a new independent method
def get_functional_eval_with_signature(train_step, mesh, state_mesh_annotations, model, config):
  ...

```

File: MaxText/multihost_dataloading.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/multihost_dataloading.py:110]: Hi @ZhiyuLi-goog , my grain changes were in this morning. I had some refactoring on tfds input pipeline as well. Looks like it may cause conflicts with your change. Let me know if you have questions / want to discuss any change

File: MaxText/train.py (hunk 3 above)
Comment (rwitten) [reply MaxText/train.py:209]: yes this makes sense.

File: MaxText/train.py (hunk 4 above)
Comment (rwitten) [reply MaxText/train.py:330]: that is a good idea and an improvement. But you've pushed the problem to:
```
return functools.partial(train_step, model, config, is_train=False)
```

File: MaxText/train.py (hunk 6 above)
Comment (rwitten) [reply MaxText/train.py:438]: I mean maxtext print!

File: MaxText/train.py (hunk 4 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py:330]: Hi, Rafi, thank you for your comment. Currently, I combined train and eval [under the same function](https://github.com/google/maxtext/blob/73665db73eb108a23e845b5d8220672d0afd6d71/MaxText/train.py#L181-L252). `is_train` flag is used to differentiate the branches. This is mainly to reuse loss_fn to ensure a compact code style.  
Looks like you are favoring more towards 2 separate functions instead. How about we do this way:
```
# we can almost reuse 99% of loss_fn for both train_step and eval_step
def loss_fn(params, is_train=True):
    logits, intermediate_outputs = model.apply({'params': params},
                         ...
                         enable_dropout=config.enable_dropout if is_train else False,  # only difference
    ...

# independent functions elsewhere.
def train_step
  ...
functional_train = get_functional_train_with_signature
p_train_step = jax.jit(functional_train)

def eval_step
  ...
functional_eval = get_functional_eval_with_signature
p_eval_step = jax.jit(functional_eval)
```

File: MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
+
```
[the remark this answers is not in the document]
Comment (ZhiyuLi-goog) [reply MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py:1]: Gotcha, @gobbleturk is convergence step appropriate metrics for comparison? I expect the packing way in MLPerf is helpful in the number of convergence since there is more meaningful tokens with less paddings.  

If we need to collect all data points for different models, it might be a long period. Not sure if we can go back to it in March after getting best gpt3 training perf.

File: MaxText/train.py (hunk 5 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py:394]: Good suggestion! Will do.

File: MaxText/train.py (hunk 6 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py:438]: Good suggestion, will add maxtext print to notify early stop.

File: MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rwitten) [reply MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py:1]: No we don't need to collect all data points for different models, just Matt's convergence sweep.

File: MaxText/train.py (hunk 4 above)
Comment (rwitten) [reply MaxText/train.py:330]: SGTM

File: MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZhiyuLi-goog) [reply MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py:1]: * [Visualization of different data format](https://colab.sandbox.google.com/drive/16Idie6590780l__4txwyLpkdSIIBZEcF#scrollTo=CJyHgzfncFUi)
* [TODO]convergence test data points

File: MaxText/train.py (hunk 4 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py]: Done.

File: MaxText/train.py (hunk 5 above)
Comment (ZhiyuLi-goog) [reply MaxText/train.py]: Disabled all mllogs in main branch and will cover it in a specific version/branch.

File: MaxText/input_pipeline/_tfds_data_processing.py [hunk 8]
```diff
@@ -187,7 +191,7 @@ def get_datasets(
   # eval_data = get_raw_dataset(eval_ds_builder, config.eval_split)
   eval_ds = eval_ds_builder.as_dataset(split=config.eval_split,
                                           read_config = read_config,
-                                          shuffle_files=config.enable_data_shuffling)
+                                          shuffle_files=False)
```
Comment (rwitten) [inline MaxText/input_pipeline/_tfds_data_processing.py:194]: Wait we want to offer this still right but set to False? am I missing something?

File: MaxText/standalone_dataloader.py [hunk 9]
```diff
@@ -38,22 +36,20 @@ def data_load_loop(config, state=None):
...
-  _, _, _, _, _, _, _, data_iterator, state = setup_train_loop(config)
+  _, _, _, _, _, _, _, data_iterator, _, state = setup_train_loop(config)
 
-  devices_array = max_utils.create_device_mesh(config)
-  mesh = Mesh(devices_array, config.mesh_axes)
   example_batch = None
 
   start = datetime.datetime.now()
   start_step = get_first_step(state)
-  example_batch = load_next_batch(data_iterator, example_batch, config, mesh)
```
Comment (rwitten) [inline MaxText/standalone_dataloader.py:49]: please talk to @aireenmei -- not sure why you all are taking and removing mesh :-)

**Review by rwitten — COMMENTED** [review]
Hold on a tiny nit, looks almost awesome.

File: MaxText/input_pipeline/_tfds_data_processing.py (hunk 8 above)
Comment (ZhiyuLi-goog) [reply MaxText/input_pipeline/_tfds_data_processing.py:194]: Thank you Rafi, if I am understanding correctly, evaluation dataset is usually not recommended to be shuffled since it won't help in back prop and it might introduce subtle differences from different orderings between 2 eval runs. Not an issue for c4_mlperf eval dataset which loads processed data directly.

I found it when checking if we can get the same first eval batch between 2 different eval runs.

File: MaxText/standalone_dataloader.py (hunk 9 above)
Comment (ZhiyuLi-goog) [reply MaxText/standalone_dataloader.py:49]: Thank you @aireenmei & @rwitten.

`load_next_batch` function is folded in [`MultiHostDataLoadIterator`](https://github.com/google/maxtext/blob/de24bde15c264704c99a12c433a14d8aa75cebed/MaxText/multihost_dataloading.py#L96) which accepts both `tf.data.dataset` and `grain.DataLoader`. `mesh` is not needed since it is part of the class. 

The main goal is to ensure repeatable evaluation loop. We may trigger eval run multiple times and need a way to [reset evaluation iterator](https://github.com/google/maxtext/blob/de24bde15c264704c99a12c433a14d8aa75cebed/MaxText/tests/tfds_data_processing_test.py#L128)

Feel free to let me know what you think.

Comment (aireenmei) [reply MaxText/standalone_dataloader.py:49]: Thanks for the explanation. LGTM

File: MaxText/input_pipeline/_tfds_data_processing.py (hunk 8 above)
Comment (rwitten) [reply MaxText/input_pipeline/_tfds_data_processing.py:194]: my mistake, thank you

File: MaxText/pyconfig.py (hunk 2 above)
Comment (rwitten) [reply MaxText/pyconfig.py:160]: @ZhiyuLi-goog I think this changed a little upstream but I have no objection.

**Review by rwitten — APPROVED** [review]
LGTM
